sql/parser/eval.go

```diff
@@ -1116,16 +1119,95 @@ func evalNotExpr(expr *NotExpr) (Datum, error) {
 	return !v, nil
 }
 
-func evalNullCheck(expr *NullCheck) (Datum, error) {
+func evalIsExpr(expr *IsExpr) (Datum, error) {
+	d, err := EvalExpr(expr.Expr)
+	if err != nil {
+		return DNull, err
+	}
+	switch expr.Operator {
+	case IsNull, IsUnknown:
+		return DBool(d == DNull), nil
+	case IsNotNull, IsNotUnknown:
+		return DBool(d != DNull), nil
+	case IsTrue:
+		return DBool(d == DBool(true)), nil
+	case IsNotTrue:
+		return DBool(d != DBool(true)), nil
+	case IsFalse:
+		return DBool(d == DBool(false)), nil
+	case IsNotFalse:
+		return DBool(d != DBool(false)), nil
+	default:
+		return DNull, util.Errorf("eval: unsupported IS operator: %d", expr.Operator)
+	}
+}
+
+func evalIsOfTypeExpr(expr *IsOfTypeExpr) (Datum, error) {
 	d, err := EvalExpr(expr.Expr)
 	if err != nil {
 		return DNull, err
 	}
-	v := d == DNull
+
+	result := DBool(true)
 	if expr.Not {
-		v = !v
+		result = !result
+	}
+
+	switch d.(type) {
+	case DBool:
+		for _, t := range expr.Types {
+			if _, ok := t.(*BoolType); ok {
+				return result, nil
+			}
+		}
+
+	case DInt:
+		for _, t := range expr.Types {
+			if _, ok := t.(*IntType); ok {
+				return result, nil
+			}
+		}
+
+	case DFloat:
+		for _, t := range expr.Types {
+			if _, ok := t.(*FloatType); ok {
+				return result, nil
+			}
+		}
+
+	case DString:
+		for _, t := range expr.Types {
+			if _, ok := t.(*StringType); ok {
+				return result, nil
+			}
+			if _, ok := t.(*BytesType); ok {
```

Comment:
We conflate them when we fill placeholders. Our handling of bytes vs strings in expressions is inadequate at the moment. We do need to keep them as distinct datum types. It has been on my mental TODO list. I'll file an issue.